File: backend/app/schemas/learning.py
"""Pydantic-схемы модуля обучения (этап 1: учётки и аутентификация learner)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

_EMAIL_MAX = 255


def _validate_email(value: str) -> str:
    # EmailStr не используем: DNS-проверка pydantic падает на .local-доменах
    # (gotcha из AGENTS.md). Минимальная структурная проверка — домен/формат
    # подтверждается отправкой письма.
    value = value.strip().lower()
    if "@" not in value or value.count("@") != 1:
        raise ValueError("Некорректный email")
    local, _, domain = value.partition("@")
    if not local or not domain or " " in value or len(domain.split(".")) < 2:
        raise ValueError("Некорректный email")
    return value


class LearningAccountCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    email: str = Field(min_length=3, max_length=_EMAIL_MAX)
    full_name: str = Field(min_length=1, max_length=255)
    department: str | None = Field(default=None, max_length=255)
    position: str | None = Field(default=None, max_length=255)

    @field_validator("email")
    @classmethod
    def _email(cls, v: str) -> str:
        return _validate_email(v)


class LearningAccountOut(BaseModel):
    """Представление учётки для админ-UI. Passwordless (миграция 113):
    парольных полей нет вовсе."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    email: str
    full_name: str
    department: str | None = None
    position: str | None = None
    status: str
    last_login_at: datetime | None = None
    created_at: datetime


class LearningAccountListOut(BaseModel):
    items: list[LearningAccountOut]
    total: int
    limit: int
    offset: int


class LearningAccountImportError(BaseModel):
    row: int = Field(ge=1)
    message: str


class LearningAccountImportOut(BaseModel):
    created: int = Field(ge=0)
    skipped_duplicates: int = Field(ge=0)
    error_count: int = Field(ge=0)
    errors: list[LearningAccountImportError]


class LearningLoginRequest(BaseModel):
    """Шаг 1 входа: запрос кода. Ответ всегда ``{"ok": true}`` — анти-enumeration."""

    model_config = ConfigDict(str_strip_whitespace=True)

    email: str = Field(min_length=3, max_length=_EMAIL_MAX)

    @field_validator("email")
    @classmethod
    def _email(cls, v: str) -> str:
        return _validate_email(v)


class LearningVerifyRequest(BaseModel):
    """Шаг 2 входа: сверка кода из письма."""

    model_config = ConfigDict(str_strip_whitespace=True)

    email: str = Field(min_length=3, max_length=_EMAIL_MAX)
    code: str = Field(min_length=6, max_length=6, pattern=r"^\d{6}$")

    @field_validator("email")
    @classmethod
    def _email(cls, v: str) -> str:
        return _validate_email(v)


class LearningLoginCodeOut(BaseModel):
    """Ручная выдача кода админом (запасной путь «письмо не дошло»).
    Plaintext-код возвращается только здесь и только админу."""

    code: str
    expires_at: datetime


# ── Курсы (инкремент 2) ──────────────────────────────────────────────────────


class CourseCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    slug: str | None = Field(default=None, max_length=140)
    # Категория курса (миграция 109): необязательная группировка на странице
    # «Обучение». Невалидный/soft-deleted id → 422 (сервис).
    category_id: uuid.UUID | None = None
    # Обязательный «для всех сотрудников» (миграция 111): курс автоматически
    # у каждого активного сотрудника, включая созданных позже.
    for_all_staff: bool = False
    # Опциональный срок прохождения (этап 2, §15): напоминание участникам — cron.
    deadline_at: datetime | None = None


class CourseUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    slug: str | None = Field(default=None, min_length=1, max_length=140)
    # Отсутствующий ключ = «не менять»; явный null = снять категорию.
    category_id: uuid.UUID | None = None
    # Наличие ключа в model_fields_set = применять (bool-поле, null невалиден).
    for_all_staff: bool | None = None
    # None = «не менять»; явная передача null (проверка model_fields_set в
    # роутере) = снять дедлайн.
    deadline_at: datetime | None = None


class CourseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    title: str
    description: str | None = None
    status: str
    category_id: uuid.UUID | None = None
    for_all_staff: bool = False
    published_at: datetime | None = None
    created_at: datetime
    deadline_at: datetime | None = None
    # Путь превью и updated_at наружу не идут (exclude) — клиент получает
    # готовый cover_url с версионным ?v= для сброса кэша браузера.
    cover_path: str | None = Field(default=None, exclude=True)
    updated_at: datetime | None = Field(default=None, exclude=True)
    cover_url: str | None = None

    @model_validator(mode="after")
    def build_cover_url(self) -> CourseOut:
        if self.cover_path:
            v = int(self.updated_at.timestamp()) if self.updated_at else 0
            self.cover_url = f"/api/v1/learning/admin/courses/{self.id}/cover?v={v}"
        return self


class CourseListOut(BaseModel):
    items: list[CourseOut]
    total: int
    limit: int
    offset: int


def _check_url_scheme(v: str | None) -> str | None:
    if v is None:
        return v
    if not v.startswith(("http://", "https://")):
        raise ValueError("Схема ссылки должна быть http(s)")
    return v


class ItemCreate(BaseModel):
    """Материал: url опционален — PDF-файл прикладывается отдельным вызовом
    к созданному элементу (§6.2); тест: только заготовка элемента.
    description — rich-описание/комментарий материала (Markdown, sanitize
    на записи в сервисе, как у описания курса)."""

    model_config = ConfigDict(str_strip_whitespace=True)

    type: str = Field(pattern="^(material|test|section)$")
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    url: str | None = Field(default=None, max_length=2048)
    file_path: str | None = Field(default=None, max_length=512)

    @field_validator("url")
    @classmethod
    def _url_scheme(cls, v: str | None) -> str | None:
        return _check_url_scheme(v)


class ItemUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str | None = Field(default=None, min_length=1, max_length=255)
    # Отсутствующий ключ = «не менять»; явный null = очистить описание.
    description: str | None = Field(default=None, max_length=8000)
    url: str | None = Field(default=None, max_length=2048)

    @field_validator("url")
    @classmethod
    def _url_scheme(cls, v: str | None) -> str | None:
        return _check_url_scheme(v)


class ItemsReorder(BaseModel):
    ordered_ids: list[uuid.UUID] = Field(min_length=1, max_length=200)


class ItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    course_id: uuid.UUID
    type: str
    title: str
    description: str | None = None
    sort_order: int
    url: str | None = None
    file_path: str | None = None


class CourseDetailOut(CourseOut):
    items: list[ItemOut] = []


class ParticipantEnroll(BaseModel):
    user_id: uuid.UUID | None = None
    learning_account_id: uuid.UUID | None = None


class ParticipantBulkEnroll(BaseModel):
    """Групповое зачисление сотрудников (этап 2, §15 «групповое зачисление
    выборкой»): дубли пропускаются, отчёт — в ответе."""

    user_ids: list[uuid.UUID] = Field(min_length=1, max_length=100)


class ParticipantOut(BaseModel):
    id: uuid.UUID
    participant_kind: str  # 'staff' | 'external'
    display_name: str
    email: str | None = None
    enrolled_at: datetime
    progress_completed: int = 0
    progress_total: int = 0
    has_certificate: bool = False


class CourseProgressOut(BaseModel):
    course_id: uuid.UUID
    total_items: int
    participants: list[ParticipantOut]


class ParticipantItemStatusOut(BaseModel):
    """Строка детализации «как решён курс» у участника (методист)."""

    item_id: uuid.UUID
    title: str
    type: str  # 'material' | 'test'
    completed: bool
    test_passed: bool = False
    attempts_submitted: int = 0


class ParticipantItemsOut(BaseModel):
    participant_id: uuid.UUID
    items: list[ParticipantItemStatusOut]


class ParticipantAttemptsResetOut(BaseModel):
    ok: bool = True
    deleted_attempts: int


# ── Тесты и попытки (инкремент 3) ────────────────────────────────────────────


class TestSettingsUpdate(BaseModel):
    """Частичная правка настроек теста. max_attempts=0 — без ограничения
    попыток; pass_score в процентах (§6.3)."""

    model_config = ConfigDict(str_strip_whitespace=True)

    pass_score: int | None = Field(default=None, ge=0, le=100)
    max_attempts: int | None = Field(default=None, ge=0, le=100)
    # Лимит времени попытки в минутах; отсутствует в теле = не менять,
    # явный null = снять ограничение (model_fields_set в роутере).
    time_limit_minutes: int | None = Field(default=None, ge=1, le=600)
    shuffle_questions: bool | None = None
    shuffle_answers: bool | None = None


class TestSettingsOut(BaseModel):
    pass_score: int
    max_attempts: int
    shuffle_questions: bool
    shuffle_answers: bool
    time_limit_minutes: int | None = None


class OptionIn(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    text: str = Field(min_length=1, max_length=2000)
    is_correct: bool = False
    sort_order: int = Field(default=0, ge=0)


class OptionOut(BaseModel):
    id: uuid.UUID
    text: str
    is_correct: bool
    sort_order: int


class QuestionCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    text: str = Field(min_length=1, max_length=8000)
    multi: bool = False
    options: list[OptionIn] = Field(min_length=2, max_length=12)

    @field_validator("options")
    @classmethod
    def _correct_options(cls, v: list[OptionIn]) -> list[OptionIn]:
        correct = sum(1 for o in v if o.is_correct)
        if correct == 0:
            raise ValueError("Нужен хотя бы один правильный вариант")
        return v


class QuestionOut(BaseModel):
    """Админское представление вопроса — с флагами правильности."""

    id: uuid.UUID
    text: str
    multi: bool
    sort_order: int
    options: list[OptionOut] = []


class ImportedQuestionIn(BaseModel):
    """Вопрос из xlsx-импорта (этап 2): шаблон фиксирует варианты A–D,
    поэтому верхний лимит вариантов жёстче, чем у ручного ввода."""

    model_config = ConfigDict(str_strip_whitespace=True)

    text: str = Field(min_length=1, max_length=8000)
    multi: bool = False
    options: list[OptionIn] = Field(min_length=2, max_length=4)

    @field_validator("options")
    @classmethod
    def _correct_options(cls, v: list[OptionIn]) -> list[OptionIn]:
        correct = sum(1 for o in v if o.is_correct)
        if correct == 0:
            raise ValueError("Нужен хотя бы один правильный вариант")
        return v


class LearnerOptionOut(BaseModel):
    """Вариант участнику: без флага правильности (ТЗ §6.3)."""

    id: uuid.UUID
    text: str


class LearnerQuestionOut(BaseModel):
    id: uuid.UUID
    text: str
    multi: bool
    options: list[LearnerOptionOut] = []


class SubmitAnswers(BaseModel):
    answers: dict[uuid.UUID, list[uuid.UUID]]


class AttemptBrief(BaseModel):
    id: uuid.UUID
    status: str  # open | submitted | abandoned
    score: int | None = None
    passed: bool | None = None
    started_at: datetime
    submitted_at: datetime | None = None


class AttemptView(BaseModel):
    """Открытая попытка: вопросы в порядке снапшота, без ответов."""

    id: uuid.UUID
    status: str
    test_item_id: uuid.UUID
    questions: list[LearnerQuestionOut]
    max_attempts: int
    submitted_count: int
    remaining_attempts: int | None = None  # None = неограниченно (max_attempts=0)
    started_at: datetime
    # Таймер попытки (этап 2): лимит в минутах и абсолютный край
    # (started_at + лимит); None = ограничение не задано.
    time_limit_minutes: int | None = None
    expires_at: datetime | None = None


class AttemptResult(BaseModel):
    """Результат попытки: балл и статус; правильные ответы не раскрываются."""

    id: uuid.UUID
    status: str
    score: int | None = None
    passed: bool | None = None
    remaining_attempts: int | None = None
    started_at: datetime
    submitted_at: datetime | None = None


class MyAttemptsOut(BaseModel):
    test_item_id: uuid.UUID
    max_attempts: int
    submitted_count: int
    remaining_attempts: int | None = None
    attempts: list[AttemptBrief]


class MyCourseItemOut(BaseModel):
    id: uuid.UUID
    type: str
    title: str
    description: str | None = None
    sort_order: int
    url: str | None = None
    has_file: bool = False
    completed: bool = False


class MyCourseOut(CourseOut):
    progress_completed: int = 0
    progress_total: int = 0
    # Денормализация категории для группировки на странице «Обучение»
    # (порядок блока — category_sort из справочника).
    category_title: str | None = None
    category_sort: int | None = None

    @model_validator(mode="after")
    def rebuild_cover_url(self) -> MyCourseOut:
        """Участнику обложка отдаётся через me-роут (проверка зачисления и
        публикации), поэтому админский URL из CourseOut перезаписывается
        (validators Pydantic v2 исполняются parent-first — итог здесь)."""
        if self.cover_path:
            v = int(self.updated_at.timestamp()) if self.updated_at else 0
            self.cover_url = f"/api/v1/learning/me/courses/{self.slug}/cover?v={v}"
        return self


class MyCourseDetailOut(MyCourseOut):
    items: list[MyCourseItemOut] = []


# ── методисты (learning_admins; назначает только глобальный админ, §3 ТЗ) ────


class LearningAdminCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    user_id: uuid.UUID


class LearningAdminOut(BaseModel):
    user_id: uuid.UUID
    full_name: str
    email: str
    added_at: datetime


# ── мета модуля (обоим контурам: портал + learn; без секретов) ───────────────


class LearningMetaOut(BaseModel):
    """Публичные параметры модуля: origin'ы, iframe с которых разрешён.

    Источник — runtime-настройка ``system.json → video_iframe_origins``
    (Admin UI → System). Гейт плеера материалов и бейджа «видео» на обоих
    контурах; learn-сборка не имеет /bootstrap, поэтому читает их здесь.
    """

    video_iframe_origins: list[str] = []


# ── Категории курсов (справочник модуля; миграция 109) ───────────────────────


class CategoryCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str = Field(min_length=1, max_length=255)


class CategoryUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str | None = Field(default=None, min_length=1, max_length=255)


class CategoryOut(BaseModel):
    id: uuid.UUID
    title: str
    sort_order: int
    # Сколько активных курсов в категории (для UI справочника).
    course_count: int = 0


class CategoryReorder(BaseModel):
    ordered_ids: list[uuid.UUID] = Field(min_length=1, max_length=200)

File: frontend/src/api/learning.ts
import { ofetch } from 'ofetch'
import { api, apiUpload, BASE_URL } from './index'
import type { components } from './types.gen'

// ── Типы из OpenAPI (backend/app/schemas/learning.py) ───────────────────────

export type LearningCourse = components['schemas']['CourseOut']
export type LearningCourseList = components['schemas']['CourseListOut']
export type LearningCourseDetail = components['schemas']['CourseDetailOut']
export type LearningCourseCreate = components['schemas']['CourseCreate']
export type LearningCourseUpdate = components['schemas']['CourseUpdate']
export type LearningItem = components['schemas']['ItemOut']
export type LearningItemCreate = components['schemas']['ItemCreate']
export type LearningItemUpdate = components['schemas']['ItemUpdate']
export type LearningTestSettingsUpdate = components['schemas']['TestSettingsUpdate']
// QuestionCreate из OpenAPI требует sort_order в вариантах — но это внутренняя
    // нормализация бэкенда (sort_order выставляется по порядку ввода), фронт его
    // не присылает. Фронтовой контракт — без sort_order.
export interface LearningOptionIn {
  text: string
  is_correct: boolean
}

export interface LearningQuestionCreate {
  text: string
  multi: boolean
  options: LearningOptionIn[]
}

// Админские view теста/вопросов: эндпоинты отдают dict (без response_model),
// поэтому этих схем в OpenAPI нет — контракт фиксируем здесь (прецедент
// HelpdeskInboxParams в api/helpdesk.ts).
export interface LearningTestSettings {
  pass_score: number
  max_attempts: number
  shuffle_questions: boolean
  shuffle_answers: boolean
}

export interface LearningOptionOut {
  id: string
  text: string
  is_correct: boolean
  sort_order: number
}

export interface LearningQuestionOut {
  id: string
  text: string
  multi: boolean
  sort_order: number
  options: LearningOptionOut[]
}
// ParticipantOut в OpenAPI занят helpdesk-схемой; learning-вариант объявляем явно.
export interface LearningParticipantRow {
  id: string
  participant_kind: 'staff' | 'external'
  display_name: string
  email: string | null
  enrolled_at: string
  progress_completed: number
  progress_total: number
  has_certificate?: boolean
}
export type LearningProgress = {
  course_id: string
  total_items: number
  participants: LearningParticipantRow[]
}
// Детализация «как решён курс» у участника (панель участников, 2026-09-01).
export interface LearningParticipantItemStatus {
  item_id: string
  title: string
  type: 'material' | 'test'
  completed: boolean
  test_passed: boolean
  attempts_submitted: number
}
export type LearningParticipantItems = {
  participant_id: string
  items: LearningParticipantItemStatus[]
}
export type LearningAccount = components['schemas']['LearningAccountOut']
export type LearningAccountList = components['schemas']['LearningAccountListOut']
export type LearningAccountCreate = components['schemas']['LearningAccountCreate']
export type LearningAccountImport = components['schemas']['LearningAccountImportOut']
export type MyCourse = components['schemas']['MyCourseOut']

/** Тестовый конфиг = настройки + вопросы с ответами (админский view). */
export interface LearningTestConfig extends LearningTestSettings {
  questions: LearningQuestionOut[]
  time_limit_minutes: number | null
}

// ── Курсы ────────────────────────────────────────────────────────────────────

export function fetchAdminCourses(params: { q?: string; limit?: number; offset?: number } = {}) {
  return api<LearningCourseList>('/learning/admin/courses', { query: params })
}

export function fetchAdminCourse(courseId: string) {
  return api<LearningCourseDetail>(`/learning/admin/courses/${courseId}`)
}

export function createCourse(body: LearningCourseCreate) {
  return api<LearningCourse>('/learning/admin/courses', { method: 'POST', body })
}

export function updateCourse(courseId: string, body: LearningCourseUpdate) {
  return api<LearningCourse>(`/learning/admin/courses/${courseId}`, { method: 'PATCH', body })
}

export function setCoursePublished(courseId: string, published: boolean) {
  return api<{ ok: boolean; status: string }>(
    `/learning/admin/courses/${courseId}/${published ? 'publish' : 'unpublish'}`,
    { method: 'POST' },
  )
}

export function deleteCourse(courseId: string) {
  return api<{ ok: boolean }>(`/learning/admin/courses/${courseId}`, { method: 'DELETE' })
}

// ── Элементы ─────────────────────────────────────────────────────────────────

export function addCourseItem(courseId: string, body: LearningItemCreate) {
  return api<{ id: string; sort_order: number }>(`/learning/admin/courses/${courseId}/items`, {
    method: 'POST',
    body,
  })
}

export function updateCourseItem(itemId: string, body: LearningItemUpdate) {
  return api<{ ok: boolean }>(`/learning/admin/courses/items/${itemId}`, {
    method: 'PATCH',
    body,
  })
}

export function deleteCourseItem(itemId: string) {
  return api<{ ok: boolean }>(`/learning/admin/courses/items/${itemId}`, { method: 'DELETE' })
}

export function reorderCourseItems(courseId: string, orderedIds: string[]) {
  return api<{ ok: boolean }>(`/learning/admin/courses/${courseId}/items/reorder`, {
    method: 'POST',
    body: { ordered_ids: orderedIds },
  })
}

export function uploadMaterialFile(itemId: string, file: File) {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<{ ok: boolean; file_path: string; mime: string }>(
    `/learning/admin/courses/items/${itemId}/file`,
    form,
  )
}

// ── Обложка курса (этап 2, ТЗ §6.2) ──────────────────────────────────────────

export function uploadCourseCover(courseId: string, file: File) {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<LearningCourse>(`/learning/admin/courses/${courseId}/cover`, form)
}

export function deleteCourseCover(courseId: string) {
  return api<LearningCourse>(`/learning/admin/courses/${courseId}/cover`, { method: 'DELETE' })
}

// ── Тесты и вопросы ──────────────────────────────────────────────────────────

export function fetchTestConfig(itemId: string) {
  return api<LearningTestConfig>(`/learning/admin/courses/items/${itemId}/test`)
}

export function updateTestSettings(itemId: string, body: LearningTestSettingsUpdate) {
  return api<{ ok: boolean }>(`/learning/admin/courses/items/${itemId}/test`, {
    method: 'PATCH',
    body,
  })
}

export function addQuestion(itemId: string, body: LearningQuestionCreate) {
  return api<{ ok: boolean; question_id: string }>(
    `/learning/admin/courses/items/${itemId}/questions`,
    { method: 'POST', body },
  )
}

export function updateQuestion(itemId: string, questionId: string, body: LearningQuestionCreate) {
  return api<{ ok: boolean }>(
    `/learning/admin/courses/items/${itemId}/questions/${questionId}`,
    { method: 'PATCH', body },
  )
}

export function deleteQuestion(itemId: string, questionId: string) {
  return api<{ ok: boolean }>(
    `/learning/admin/courses/items/${itemId}/questions/${questionId}`,
    { method: 'DELETE' },
  )
}

// ── Участники и прогресс ─────────────────────────────────────────────────────

export function enrollParticipant(
  courseId: string,
  body: components['schemas']['ParticipantEnroll'],
) {
  return api<{ ok: boolean; participant_id: string }>(
    `/learning/admin/courses/${courseId}/participants`,
    { method: 'POST', body },
  )
}

/** Групповое зачисление сотрудников (этап 2): дубли пропускаются. */
export function bulkEnrollParticipants(courseId: string, userIds: string[]) {
  return api<{ enrolled: number; skipped_duplicates: number; errors: { user_id: string; message: string }[] }>(
    `/learning/admin/courses/${courseId}/participants/bulk`,
    { method: 'POST', body: { user_ids: userIds } },
  )
}

export function unenrollParticipant(courseId: string, participantId: string) {
  return api<{ ok: boolean }>(
    `/learning/admin/courses/${courseId}/participants/${participantId}`,
    { method: 'DELETE' },
  )
}

/** Статус решения курса участником: каждый элемент с флагом и попытками. */
export function fetchParticipantItems(courseId: string, participantId: string) {
  return api<LearningParticipantItems>(
    `/learning/admin/courses/${courseId}/participants/${participantId}/items`,
  )
}

/** Сброс попыток теста участнику: попытки + отметка «пройдено» (аудит на бэке). */
export function resetParticipantAttempts(courseId: string, participantId: string, itemId: string) {
  return api<{ ok: boolean; deleted_attempts: number }>(
    `/learning/admin/courses/${courseId}/participants/${participantId}/items/${itemId}/reset-attempts`,
    { method: 'POST' },
  )
}

/** Сертификат участника (PDF): скачивание выданного либо выпуск за пройденный курс. */
export function fetchParticipantCertificateBlob(courseId: string, participantId: string) {
  return apiBlob(
    `/learning/admin/courses/${courseId}/participants/${participantId}/certificate`,
  )
}


export function fetchCourseProgress(courseId: string) {
  return api<LearningProgress>(`/learning/admin/courses/${courseId}/progress`)
}

/** Blob-запрос (скачивание xlsx/pdf) с cookie-сессией портала. */
function apiBlob(path: string): Promise<Blob> {
  return ofetch(path, { baseURL: BASE_URL, credentials: 'include', responseType: 'blob' }) as Promise<Blob>
}

/** Экспорт прогресса курса в xlsx (этап 2): blob для скачивания. */
export function exportProgressBlob(courseId: string) {
  return apiBlob(`/learning/admin/courses/${courseId}/progress/export`)
}

// ── Внешние учётки ───────────────────────────────────────────────────────────

// ── Категории курсов (справочник модуля, миграция 109) ───────────────────────

export type LearningCategory = components['schemas']['CategoryOut']
export type LearningCategoryCreate = components['schemas']['CategoryCreate']
export type LearningCategoryUpdate = components['schemas']['CategoryUpdate']
export type LearningCategoryReorder = components['schemas']['CategoryReorder']

export function fetchCategories() {
  return api<LearningCategory[]>('/learning/admin/categories')
}

export function createCategory(body: LearningCategoryCreate) {
  return api<LearningCategory>('/learning/admin/categories', { method: 'POST', body })
}

export function updateCategory(id: string, body: LearningCategoryUpdate) {
  return api<LearningCategory>(`/learning/admin/categories/${id}`, { method: 'PATCH', body })
}

export function reorderCategories(body: LearningCategoryReorder) {
  return api<{ ok: boolean }>('/learning/admin/categories/reorder', { method: 'POST', body })
}

export function deleteCategory(id: string) {
  return api<{ ok: boolean }>(`/learning/admin/categories/${id}`, { method: 'DELETE' })
}

export function fetchLearningAccounts(
  params: { q?: string; limit?: number; offset?: number; status?: string } = {},
) {
  return api<LearningAccountList>('/learning/admin/accounts', { query: params })
}

export function createLearningAccount(body: LearningAccountCreate) {
  return api<LearningAccount>('/learning/admin/accounts', { method: 'POST', body })
}

export function importLearningAccounts(file: File) {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<LearningAccountImport>('/learning/admin/accounts/import', form)
}

export function downloadLearningAccountsTemplate() {
  return apiBlob('/learning/admin/accounts/template')
}

// ── Импорт вопросов из xlsx (этап 2, ТЗ §13/§15) ─────────────────────────────

export interface QuestionsImportRow {
  row: number
  text: string
  multi: boolean
  options: { text: string; is_correct: boolean }[]
}

export interface QuestionsImportError {
  row: number
  message: string
}

export interface QuestionsImportPreview {
  questions: QuestionsImportRow[]
  errors: QuestionsImportError[]
}

export interface QuestionsImportResult {
  created: number
  errors: QuestionsImportError[]
}

/** Шаблон импорта вопросов: колонки = тем же кодом, что валидирует импорт. */
export function downloadQuestionsTemplate(itemId: string) {
  return apiBlob(`/learning/admin/courses/items/${itemId}/questions/template`)
}

export function previewQuestionsImport(itemId: string, file: File) {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<QuestionsImportPreview>(
    `/learning/admin/courses/items/${itemId}/questions/import/preview`,
    form,
  )
}

export function importQuestions(itemId: string, file: File) {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<QuestionsImportResult>(
    `/learning/admin/courses/items/${itemId}/questions/import`,
    form,
  )
}

/** Ручная выдача одноразового кода входа (запасной путь «письмо не дошло»). */
export function issueLearningAccountLoginCode(accountId: string) {
  return api<{ code: string; expires_at: string }>(
    `/learning/admin/accounts/${accountId}/login-code`,
    { method: 'POST' },
  )
}

export function blockLearningAccount(accountId: string) {
  return api<{ ok: boolean }>(`/learning/admin/accounts/${accountId}/block`, { method: 'POST' })
}

export function unblockLearningAccount(accountId: string) {
  return api<{ ok: boolean }>(`/learning/admin/accounts/${accountId}/unblock`, { method: 'POST' })
}

// ── Сотрудник/learner: мои курсы (виджет главной) ────────────────────────────

export function fetchMyCourses() {
  return api<MyCourse[]>('/learning/me/courses')
}

// ── Мета модуля (обоим контурам; гейт видео-плеера, без секретов) ────────────

export type LearningMeta = components['schemas']['LearningMetaOut']

export function fetchLearningMeta() {
  return api<LearningMeta>('/learning/meta')
}

// ── Learner: раздел «Обучение» (портал, инкремент learner-страниц) ───────────

export type MyCourseDetail = components['schemas']['MyCourseDetailOut']
export type MyCourseItem = components['schemas']['MyCourseItemOut']
export type AttemptView = components['schemas']['AttemptView']
export type AttemptResult = components['schemas']['AttemptResult']
export type AttemptBrief = components['schemas']['AttemptBrief']
export type MyAttempts = components['schemas']['MyAttemptsOut']
export type LearnerQuestion = components['schemas']['LearnerQuestionOut']
export type LearnerOption = components['schemas']['LearnerOptionOut']

export function fetchMyCourse(slug: string) {
  return api<MyCourseDetail>(`/learning/me/courses/${slug}`)
}

export function completeMaterial(itemId: string) {
  return api<{ ok: boolean; already_completed: boolean }>(
    `/learning/me/items/${itemId}/complete`,
    { method: 'POST' },
  )
}

/** PDF материала: стриминг, cookie портальной сессии — обычная ссылка. */
export function materialFileUrl(itemId: string) {
  return `/api/v1/learning/me/items/${itemId}/file`
}

/** PDF сертификата за пройденный курс: та же схема, что у материала. */
export function certificateFileUrl(slug: string) {
  return `/api/v1/learning/me/courses/${slug}/certificate`
}

export function fetchMyAttempts(testItemId: string) {
  return api<MyAttempts>(`/learning/me/tests/${testItemId}/my-attempts`)
}

export function startAttempt(testItemId: string) {
  return api<AttemptView>(`/learning/me/tests/${testItemId}/attempts`, { method: 'POST' })
}

export function fetchAttempt(attemptId: string) {
  return api<AttemptView | AttemptResult>(`/learning/me/attempts/${attemptId}`)
}

export function submitAttempt(attemptId: string, answers: Record<string, string[]>) {
  return api<AttemptResult>(`/learning/me/attempts/${attemptId}/submit`, {
    method: 'POST',
    body: { answers },
  })
}

// ── Методисты (learning_admins; назначает только глобальный админ, §3 ТЗ) ────

export type LearningAdmin = components['schemas']['LearningAdminOut']
export type LearningAdminCreate = components['schemas']['LearningAdminCreate']

export function fetchLearningAdmins() {
  return api<LearningAdmin[]>('/learning/admins')
}

export function assignLearningAdmin(body: LearningAdminCreate) {
  return api<{ ok: boolean; user_id: string }>('/learning/admins', { method: 'POST', body })
}

export function revokeLearningAdmin(userId: string) {
  return api<{ ok: boolean }>(`/learning/admins/${userId}`, { method: 'DELETE' })
}
